### ISSUE TITLE
Support Cleartext plugin

### ISSUE THREAD

**Issue by silverbullettruck2001** [cweijan/vscode-mysql#465]
Can this extension add a login feature to allow using the mySQL Clear text plugin?  I use this for connecting to Amazon RDS instances and it uses tokens for logging into it.  This requires using the clear text plug-in.

**Comment by cweijan** [comment]
I don't understand what plugin you are using, you should give me an code example.

**Comment by silverbullettruck2001** [comment]
@cweijan sorry for not responding sooner.  Here's the plugin I was talking about:
[link to cleartext mysql plugin documentation](https://dev.mysql.com/doc/refman/5.7/en/cleartext-pluggable-authentication.html)

I hope this helps, please let me know if I can provide more information.  Thanks!

**Comment by cweijan** [comment]
Oh, looks like it can be supported. sidorares/node-mysql2/issues/1175

**Comment by cweijan** [comment]
I added the authentication plugin according to the mysql-node2 library in version 4.6.0, but it doesn't seem to work, you can try to connect.

**Comment by silverbullettruck2001** [comment]
@cweijan it looks like the PR that adds the `mysql_clear_password auth plugin` is not yet merged.  Maybe that's why it's not working?  I posted a comment on the PR to see if it could be merged.  Here's the PR I am referencing that hasn't been merged:
[mysql-node2 library mysql_clear_password PR](https://github.com/sidorares/node-mysql2/pull/1403)

**Comment by cweijan** [comment]
This problem doesn't seem to be fixed, are you sure amazon rds uses the clear password plugin?

**Comment by silverbullettruck2001** [comment]
Yes, the instance I am connecting to uses it.  I think the issue is that
the `node-mysql2` project doesn't support it yet.  I think the developer is
still trying to get it to allow for it.  The original issue from that
project that you provided in a previous email is still pending.  Maybe you
want to look into that and see if you can resolve the issue?

**Comment by cweijan** [comment]
Maybe you should try the default way to connect to amazon rds, clear_password is an old authentication way, I don't have much will to support it.

**Comment by silverbullettruck2001** [comment]
Our administrators won't let us use any other method or else I would
definitely go that route.  Hopefully the `node-mysql2` project developer
will be able to get this fixed and it should work correctly.

**Comment by silverbullettruck2001** [comment]
@cweijan I know this is old, but I wanted to check with you.  You had mentioned previously that you implemented this in 4.6.0 and I could test this out, but I could never find the configuration to do this.  Can you provide screenshots?  Also, I have been working with the developer for `node-mysql2` and they are expecting to provide a release soon that will have this functionality available to allow for the cleartext plugin.  So hopefully once that is released you can update to the latest release of `node-mysql2` and this will work.

**Comment by silverbullettruck2001** [comment]
I should have mentioned, the PR for the change to `node-mysql2` is below.
[https://github.com/sidorares/node-mysql2/pull/1552](https://github.com/sidorares/node-mysql2/pull/1552)

**Comment by cweijan** [comment]
Thank you sir, I will update it in the next version!

**Comment by cweijan** [comment]
I've added mysql_clear_password by default in my code a long time ago, but that doesn't seem to work for you.


Is the pr you submitted on node_mysql2 able to connect to your mysql using mysql_clear_password?


I did some research before and found that the mysql server doesn't seem to tell you that mysql_clear_password is being used, which is the reason for not being able to connect using mysql_clear_password.

**Comment by silverbullettruck2001** [comment]
Yes, the PR I submitted does work using mysql_clear_password.  There is nothing in the server side that tells you it is being used, it is only present in the client connection.  Essentially you enable the authPlugin when you build the connection and node-mysql2 will transmit the password value in cleartext to the server rather than using any default hashing/encryption.  You would definitely want to provide a switch to enable this at the connection level though.  

I really appreciate your assistance with getting this added!

**Comment by cweijan** [comment]
Hi sir,  in the previous version, I followed the author's instructions for mysql2 and added authPlugins, but it doesn't seem to work for you, please try how to make the connection support it, reference the code below.

```js
const mysql = require('mysql2');


// create the connection to database
const connection = mysql.createConnection({
    host: 'localhost',
    user: 'root',
    password: 'pass',
    database: 'test',
    authPlugins: {
        mysql_clear_password: () => () => {
            return Buffer.from(node.password + '\0')
        }
    },
});

// simple query
connection.query('SELECT 1', (err, results, fields) => {
    console.log(err, results);
});

```

**Comment by silverbullettruck2001** [comment]
I don't think the change to support clear text hasn't been included in a release of `node-myql2` yet, so you will have to wait until it's released or build your own version of it based on the latest code.

**Comment by cweijan** [comment]
I checked your pr, you have added the mysql_clear_password plugin in the code, which means that the authPlugins provided by the author does not work at present version?

**Comment by silverbullettruck2001** [comment]
Correct, the PR is merged, but the author hasn't made a release of
`node-mysql2` since November 2021.  He is working on generating the release
now, but hasn't reported his progress yet.  I am hoping he will be able to
release the code in the next few days.  I have been keeping an eye on it,
so I can definitely let you know as soon as it is available, if you want?

**Comment by cweijan** [comment]
Thank's for informing.

**Comment by silverbullettruck2001** [comment]
@cweijan the developer for `node-mysql2` has created a release candidate for the build that will include support for cleartext.  It is on `npm` here [3.0.0-rc.1](https://www.npmjs.com/package/mysql2/v/3.0.0-rc.1).  He did it as a release candidate because he is still working out some kinks with the automated build process.  The final release will be the same, it will just be generated by the build process completely.  If you could do a build with this release of `node-mysql2` and see if you have any issues, then I could test out connecting using the cleartext feature.

**Comment by cweijan** [comment]
Thanks, I will try to use it, is there anything else I need to do besides upgrading the version?

**Comment by silverbullettruck2001** [comment]
I don't believe so as it should just use the code you previously provided
that utilizes the new authPlugin.  So just make sure that you expose that
in the UI somehow so that you can specify the plugin when building the
connection.

**Comment by cweijan** [comment]
I tried 3.0.0-rc.1, now the authPlugins parameter is removed, can you provide me an example of the connection?

**Comment by silverbullettruck2001** [comment]
@cweijan this appears to be a bug with the release candidate.  If I look at the code in the repo the `AuthPlugins` is there but if I look at what gets installed when doing `npm i` it does not exist.  I will report a bug and see what can be done to fix this.  Thanks!

**Comment by cweijan** [comment]
Please try the code below and change the configuration to your mysql using clear_text, if it works then it's fine.
```js
const mysql = require('mysql2');


// create the connection to database
const connection = mysql.createConnection({
    host: 'localhost',
    user: 'root',
    password: 'pass',
    database: 'test',
    authPlugins: {
        mysql_clear_password: () => () => {
            return Buffer.from(node.password + '\0')
        }
    },
});

// simple query
connection.query('SELECT 1', (err, results, fields) => {
    console.log(err, results);
});

```

**Comment by cweijan** [comment]
I have released version 2.9.0 and have upgraded node-mysql2 to the latest version.

**Comment by silverbullettruck2001** [comment]
I updated my extension in VSCode but I don't see an option in the Add/Edit Connection screen to enable the plugin.  How are you controlling this?

**Comment by cweijan** [comment]
You didn't tell me how to enable it, I kept the auth_plugin parameter and upgraded node-mysql2 to the latest version, if it doesn't work, it means your pull_request doesn't work.

**Comment by cweijan** [comment]
Hi, are you still using this extension? I created an aws mysql. By checking SSL, I was able to successfully connect using IAM Token. If you are still using this extension, please try it.

<img width="1041" alt="image" src="https://github.com/cweijan/vscode-database-client/assets/27798227/84aef650-4108-4114-829d-b10300cdc3af">

**Comment by cweijan** [comment]
I have tested that the mysql_clear_password authentication code will be triggered after enabling SSL, which means that this issue will be resolved when the 2022.12.8 version is released.

**Comment by rfcdejong** [comment]
Same for Azure with Entra Id password, must be clear-text
https://learn.microsoft.com/en-us/azure/mysql/flexible-server/how-to-azure-ad#important-considerations-when-connecting
